src/audio_decoder_midi.cpp

```diff
@@ -0,0 +1,405 @@
+	AudioDecoderMidi* midi = reinterpret_cast<AudioDecoderMidi*>(instance);
+
+	if (midi->file_buffer_pos >= midi->file_buffer.size()) {
+		return EOF;
+	}
+
+	return midi->file_buffer[midi->file_buffer_pos++];
+}
+
+bool AudioDecoderMidi::Open(Filesystem_Stream::InputStream stream) {
+	Reset();
+	seq->clear();
+
+	file_buffer_pos = 0;
+	file_buffer = Utils::ReadStream(stream);
+
+	if (!seq->load(this, read_func)) {
+		error_message = "Midi: Error reading file";
+		return false;
+	}
+	seq->rewind();
+	mtime = seq->get_start_skipping_silence();
+
+	if (!mididec->SupportsMidiMessages()) {
+		if (!mididec->Open(file_buffer)) {
+			error_message = "Internal Midi: Error reading file";
+			return false;
+		}
+	}
+
+	tempo.emplace_back(this, midi_default_tempo);
+
+	return true;
+}
+
+void AudioDecoderMidi::Pause() {
+	paused = true;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, 0);
+		mididec->SendMidiMessage(msg);
+	}
+}
+
+void AudioDecoderMidi::Resume() {
+	paused = false;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		mididec->SendMidiMessage(msg);
+	}
+}
+
+int AudioDecoderMidi::GetVolume() const {
+	if (!mididec->SupportsMidiMessages()) {
+		if (fade_steps > 0) {
+			return static_cast<int>(fade_volume_end * 100);
+		}
+		return static_cast<int>(volume * 100);
+	}
+
+	// Lie about the volume as this is handled by the Midi messages internally
+	return 100;
+}
+
+void AudioDecoderMidi::SetVolume(int new_volume) {
+	// cancel any pending fades
+	fade_steps = 0;
+
+	volume = new_volume / 100.0f;
+	for (int i = 0; i < 16; i++) {
+		uint32_t msg = midimsg_volume(i, static_cast<uint8_t>(channel_volumes[i] * volume));
+		mididec->SendMidiMessage(msg);
+	}
+}
+
+void AudioDecoderMidi::SetFade(int begin, int end, std::chrono::milliseconds duration) {
+	fade_steps = 0;
+	last_fade_mtime = 0us;
+
+	if (duration <= 0ms) {
+		SetVolume(end);
+		return;
+	}
+
+	if (begin == end) {
+		SetVolume(end);
+		return;
+	}
+
+	volume = begin / 100.0f;
+	fade_volume_end = end / 100.0f;
+	fade_steps = duration.count() / 100;
+	delta_volume_step = (fade_volume_end - volume) / fade_steps;
+}
+
+bool AudioDecoderMidi::Seek(std::streamoff offset, std::ios_base::seekdir origin) {
+	assert(!tempo.empty());
+
+	if (offset == 0 && origin == std::ios_base::beg) {
+		mtime = seq->rewind_to_loop()->time;
+
+		// When the loop points to the end of the track keep it alive to match
+		// RPG_RT behaviour.
+		loops_to_end = mtime >= seq->get_total_time();
+
+		if (mtime > 0us) {
+			// Throw away all tempo data after the loop point
+			auto rit = std::find_if(tempo.rbegin(), tempo.rend(), [&](auto& t) { return t.mtime <= mtime; });
+			auto it = rit.base();
+			if (it != tempo.end()) {
+				tempo.erase(it, tempo.end());
+			}
+
+			// Bit of a hack, prevent stuck notes
+			// TODO: verify with a MIDI event stream inspector whether RPG_RT does this?
+			// FIXME synth->all_note_off();
```

Comment:
Is this already checked and fixed with sysex gm reset or similar somewhere else?

Reply:
Turned out to be unnecessary. This is testing my memory a bit since it's now been almost a whole year, but RPG_RT does nothing of the sort, and the stuck notes I was experiencing (if i remember correctly) was mostly because the initial implementation skipped past messages that had the _same time code_ as the loopstart.